src/artifacts-cred.py

```diff
@@ -0,0 +1,177 @@
+
+
+    def _get_credentials_from_credential_provider(self, url, is_retry):
+        non_interactive = self._NON_INTERACTIVE_VAR_NAME in os.environ and \
+            os.environ[self._NON_INTERACTIVE_VAR_NAME] and \
+            str(os.environ[self._NON_INTERACTIVE_VAR_NAME]).lower() == "true"
+
+        proc = Popen(
+            self.exe + [
+                "-Uri", url,
+                "-IsRetry", str(is_retry),
+                "-NonInteractive", str(non_interactive),
+                "-CanShowDialog", "False",
+                "-OutputFormat", "Json"
+            ],
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE
+        )
+
+        # Read all standard error first, which may either display
+        # errors from the credential provider or instructions
+        # from it for Device Flow authentication.
+        for stderr_line in iter(proc.stderr.readline, b''):
+            line = stderr_line.decode("utf-8", "ignore")
+            sys.stdout.write(line)
+            sys.stdout.flush()
+
+        proc.wait()
+
+        if proc.returncode != 0:
+            stderr = proc.stderr.read().decode("utf-8", "ignore")
+            raise RuntimeError("Failed to get credentials: process with PID {pid} exited with code {code}; additional error message: {error}"
+                .format(pid=proc.pid, code=proc.returncode, error=stderr))
+
+        try:
+            # stdout is expected to be UTF-8 encoded JSON, so decoding errors are not ignored here.
+            payload = proc.stdout.read().decode("utf-8")
+        except ValueError:
+            raise RuntimeError("Failed to get credentials: the Credential Provider's output could not be decoded using UTF-8.")
+
+        try:
+            parsed = json.loads(payload)
+            return parsed["Username"], parsed["Password"]
+        except ValueError:
+            raise RuntimeError("Failed to get credentials: the Credential Provider's output could not be parsed as JSON.")
+
+class ArtifactsKeyringBackend():
+    SUPPORTED_NETLOC = (
+        "pkgs.dev.azure.com",
+        "pkgs.visualstudio.com",
+        "pkgs.codedev.ms",
+        "pkgs.vsts.me"
+    )
+    _PROVIDER = CredentialProvider
+
+    priority = 9.9
+
+    # I dont see a need to cache anything here
+    def get_credential(self, service, username):
+        try:
+            parsed = urlsplit(service)
+        except Exception as exc:
+            warnings.warn(str(exc))
+            return None
+
+        netloc = parsed.netloc.rpartition("@")[-1]
+
+        if netloc is None or not netloc.endswith(self.SUPPORTED_NETLOC):
+            return None
+
+        provider = self._PROVIDER()
+
+        username, password = provider.get_credentials(service)
+
+        if username and password:
+            return password
+
+result = sys.stdin.read()
+jsonResult = json.loads(result)
+resultUrl = "https://" + jsonResult['channel_alias']['location']
```

Comment:
Validate that `jsonResult['channel_alias']['scheme']` is https, and/or
```suggestion
resultUrl = jsonResult['channel_alias']['scheme'] + "://" + jsonResult['channel_alias']['location']
```